### PR TITLE
Secure admin services using basic authentication

### PR DESCRIPTION
**Pull request by warunalakshitha**
Broker rest services need to be secured using basic authentication. This can be done adding BasicAuthSecurityInterceptor to msf4j runner.
This also refactors the test cases of rest services to support basic auth.

Resolves https://github.com/wso2/message-broker/issues/203

## Purpose
Secure broker rest services using basic auth

## Goals
Basic auth support for rest services

## Approach
Add BasicAuthSecurityInterceptor to msf4j runner as globa request interceptor.

## Release note
Secure admin services using basic authentication

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > This will be covered with rest api tests

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? yes
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

### CODE REVIEW

File: modules/broker-auth/src/main/java/org/wso2/broker/auth/authentication/authenticator/Authenticator.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+public interface Authenticator {
+
+    /**
+     * Initiate authenticator with startup context.
+     *
+     * @param startupContext the startup context provides registered services for authenticator functionality.
+     * @param authenticationConfiguration the authentication configuration
+     */
+    void initialize(StartupContext startupContext,
+                    BrokerAuthConfiguration.AuthenticationConfiguration authenticationConfiguration);
```
Comment (a5anka) [inline modules/broker-auth/src/main/java/org/wso2/broker/auth/authentication/authenticator/Authenticator.java]: Isn't it enough to give the startup context? The implementer can get the required configurations using the startup context.

File: modules/broker-auth/src/main/java/org/wso2/broker/auth/authentication/authenticator/AuthenticatorFactory.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+     * Provides an instance of @{@link Authenticator}
+     *
+     * @param startupContext the startup context provides registered services for authenticator functionality.
+     * @param authenticationConfiguration the authentication configuration
+     * @return authenticator for given configuration
+     * @throws Exception throws if error occurred while providing new instance of authenticator
+     */
+    public Authenticator getAuthenticator(StartupContext startupContext,
+                                          BrokerAuthConfiguration.AuthenticationConfiguration
+                                                  authenticationConfiguration) throws Exception {
```
Comment (a5anka) [inline modules/broker-auth/src/main/java/org/wso2/broker/auth/authentication/authenticator/AuthenticatorFactory.java]: Do we need the `throws Exception` here?

File: modules/broker-auth/src/main/java/org/wso2/broker/auth/authentication/authenticator/impl/JaasAuthenticator.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+/**
+ * An implementation of @{@link Authenticator} for provide authentication using
+ * Jaas @{@link javax.security.auth.spi.LoginModule}.
+ */
+public class JaasAuthenticator implements Authenticator {
+
+    @Override
+    public void initialize(StartupContext startupContext,
+                           BrokerAuthConfiguration.AuthenticationConfiguration authenticationConfiguration) {
+        UserStoreManager userStoreManager = startupContext.getService(UserStoreManager.class);
```
Comment (a5anka) [inline modules/broker-auth/src/main/java/org/wso2/broker/auth/authentication/authenticator/impl/JaasAuthenticator.java]: We need to externalize the UserStoreManager class as well. Maybe in a later release. Currently, this is done in Main class. We need to move the initialization of UserStoreManager out of Main class.

File: modules/broker-auth/src/main/java/org/wso2/broker/auth/authentication/authenticator/Authenticator.java (hunk 1 above)
Comment (warunalakshitha) [reply modules/broker-auth/src/main/java/org/wso2/broker/auth/authentication/authenticator/Authenticator.java]: yes. I Generic case we should only have start up context.

File: modules/broker-auth/src/main/java/org/wso2/broker/auth/authentication/authenticator/impl/JaasAuthenticator.java (hunk 3 above)
Comment (warunalakshitha) [reply modules/broker-auth/src/main/java/org/wso2/broker/auth/authentication/authenticator/impl/JaasAuthenticator.java]: I have created a https://github.com/wso2/message-broker/issues/232 to track this.

**Pull request merged**